### PR TITLE
ref: refactor SentryRunLoopDelayTracker API and internals

### LINKED ISSUE #8288: ref: refactor SentryRunLoopDelayTracker API and internals
> [!NOTE]
> The pull request "[ref: refactor SentryRunLoopDelayTracker API and internals](https://github.com/getsentry/sentry-cocoa/pull/8287)" was created by @philprime but did not reference an issue. Therefore this issue was created for better visibility in external tools like Linear.

Extracted from https://github.com/getsentry/sentry-cocoa/pull/8179

- Renames `addOngoingHangObserver` to `addObserver` and parameter `id` to `token`
- Renames `hangNotifyThreshold` to `pollingInterval`
- Replaces `NSRecursiveLock` with `SentryMutex`
- Reorganizes class layout with MARK sections and converts comments to doc comments
- Adds tests for multi-observer edge cases (add/remove during active hang, double-remove)

#skip-changelog

### PR DESCRIPTION
**Pull request by philprime**
Extracted from https://github.com/getsentry/sentry-cocoa/pull/8179

- Renames `addOngoingHangObserver` to `addObserver` and parameter `id` to `token`
- Renames `hangNotifyThreshold` to `pollingInterval`
- Replaces `NSRecursiveLock` with `SentryMutex`
- Reorganizes class layout with MARK sections and converts comments to doc comments
- Adds tests for multi-observer edge cases (add/remove during active hang, double-remove)

#skip-changelog

Closes #8288

### CODE REVIEW

**Comment by philprime** [comment]
* **#8290** <a href="https://app.graphite.com/github/pr/getsentry/sentry-cocoa/8290?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8287** <a href="https://app.graphite.com/github/pr/getsentry/sentry-cocoa/8287?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/getsentry/sentry-cocoa/8287?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8284** <a href="https://app.graphite.com/github/pr/getsentry/sentry-cocoa/8284?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8282** <a href="https://app.graphite.com/github/pr/getsentry/sentry-cocoa/8282?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: Tests/SentryTests/Integrations/AppHangTracking/SentryRunLoopDelayTrackerTests.swift [hunk 1]
```diff
@@ -359,6 +357,150 @@ final class SentryRunLoopDelayTrackerTests: XCTestCase {
...
+
+        var remainingInterval: TimeInterval = 0
+        var remainingOngoing = false
+        let remainingExpectation = XCTestExpectation(description: "Remaining observer called")
+        let token2 = sut.addObserver { delay in
+            if remainingInterval == 0 {
+                remainingExpectation.fulfill()
+            }
+            remainingInterval = delay.duration
+            remainingOngoing = delay.isOngoing
```
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/AppHangTracking/SentryRunLoopDelayTrackerTests.swift]: `m`: I think we have to update the remainingInterval and the remainingOngoing before fulfilling the expectation, because otherwise we have a race condition
```suggestion
            remainingInterval = delay.duration
            remainingOngoing = delay.isOngoing
            if remainingInterval == 0 {
                remainingExpectation.fulfill()
            }

```

File: Tests/SentryTests/Integrations/AppHangTracking/SentryRunLoopDelayTrackerTests.swift [hunk 2]
```diff
@@ -359,6 +357,150 @@ final class SentryRunLoopDelayTrackerTests: XCTestCase {
...
+        let sut = SentryDefaultRunLoopDelayTracker(
+            dependencies: mockDependencies,
+            createObserver: createObserver,
+            addObserver: addObserver,
+            removeObserver: removeObserver,
+            queue: queue)
+
+        var removedObserverCalled = false
+        let token1 = sut.addObserver { _ in
+            removedObserverCalled = true
```
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/AppHangTracking/SentryRunLoopDelayTrackerTests.swift]: `l`: I think it would be safer to use a XCTestExcpecation with `isInverted` to avoid potential race conditions. The bool `removedObserverCalled` isn't thread safe.

File: Sources/Swift/Integrations/AppHangTracking/SentryRunLoopDelayTracker.swift [hunk 3]
```diff
@@ -70,79 +99,55 @@ final class SentryDefaultRunLoopDelayTracker<T: SentryRunLoopObserver, Dependenc
...
-        // Modifying observers requires holding the lock
-        observersLock.synchronized {
-            _ = observers.removeValue(forKey: id)
+    /// Removes the observer with the given token
+    ///
+    /// - Precondition: Must be called on main queue
+    func removeObserver(token: SentryRunLoopDelayTrackerObserverToken) {
+        // Keep a reference to the removed observer so it's destroyed after leaving the critical section.
+        let (removed, isEmpty) = observers.withLock {
+            ($0.removeValue(forKey: token), $0.isEmpty)
```
Comment (philipphofmann) [inline Sources/Swift/Integrations/AppHangTracking/SentryRunLoopDelayTracker.swift]: `m`: I'm curious why it can only be destroyed after leaving the critical section, and what exactly is the critical section?

File: Sources/Swift/Integrations/AppHangTracking/SentryRunLoopDelayTracker.swift [hunk 4]
```diff
@@ -161,61 +166,51 @@ final class SentryDefaultRunLoopDelayTracker<T: SentryRunLoopObserver, Dependenc
...
-                // Accessing observers off the main queue requires holding the lock
-                // because the main queue could be modifying it
-                observersLock.synchronized {
-                    observers.values.forEach {
-                        $0(.init(duration: duration, isOngoing: true))
-                    }
-                }
+                // Snapshot handlers outside the critical section to avoid deadlocks if a handler calls addObserver/removeObserver.
+                let handlers = observers.withLock { Array($0.values) }
+                handlers.forEach { $0(.init(duration: duration, isOngoing: true)) }
```
Comment (philipphofmann) [inline Sources/Swift/Integrations/AppHangTracking/SentryRunLoopDelayTracker.swift:208]: That minimizes the lock duration. Excellent thanks 💯

**Review by philipphofmann — COMMENTED** [review]
Thanks, almost LGTM

File: Sources/Swift/Integrations/AppHangTracking/SentryRunLoopDelayTracker.swift (hunk 3 above)
Comment (philprime) [reply Sources/Swift/Integrations/AppHangTracking/SentryRunLoopDelayTracker.swift]: The critical section is the time between acquiring the lock and releasing it. 
Generally we want to keep the time a lock is held to a minimum, but as the observers could have side-effects on deinit (e.g. an object is removing itself from NotificationCenter as an observer) we most likely don't want to be in the locked state at that point. That's why assigning the removed block to a temporary variable will keep the reference around until at least after the unlock.

Comment (philprime) [reply Sources/Swift/Integrations/AppHangTracking/SentryRunLoopDelayTracker.swift]: If the closure's deinit chain re-enters the same lock, we'd deadlock.

By capturing `removed` in the tuple return and keeping it alive until after `withLock` returns, the deallocation happens outside the lock. 

Updated the comment to make this clearer.

File: Tests/SentryTests/Integrations/AppHangTracking/SentryRunLoopDelayTrackerTests.swift (hunk 2 above)
Comment (philprime) [reply Tests/SentryTests/Integrations/AppHangTracking/SentryRunLoopDelayTrackerTests.swift]: Good call — replaced the bool with an inverted `XCTestExpectation` to make it thread-safe.

File: Tests/SentryTests/Integrations/AppHangTracking/SentryRunLoopDelayTrackerTests.swift (hunk 1 above)
Comment (philprime) [reply Tests/SentryTests/Integrations/AppHangTracking/SentryRunLoopDelayTrackerTests.swift]: Fixed — values are now updated before fulfilling the expectation. Also added `assertForOverFulfill = false` since the handler can be called multiple times (on each polling timeout).

File: Sources/Swift/Integrations/AppHangTracking/SentryRunLoopDelayTracker.swift (hunk 3 above)
Comment (philipphofmann) [reply Sources/Swift/Integrations/AppHangTracking/SentryRunLoopDelayTracker.swift]: Thanks for explaining, @philprime. It's clearer now. While I try to avoid acquiring locks in deinit, you still could have use cases for that. Can you please elaborate on what drove you to add this safety check? Did you see that problem already, or is it a precaution? I'm only curious.

**Review by philipphofmann — APPROVED** [review]
LGTM

File: Sources/Swift/Integrations/AppHangTracking/SentryRunLoopDelayTracker.swift (hunk 3 above)
Comment (philprime) [reply Sources/Swift/Integrations/AppHangTracking/SentryRunLoopDelayTracker.swift]: It's a general precaution raised during review of one of the review bots. As the handler is arbitrary there is no contract in place that it's save, and it's even reproducible in unit tests in multi-threading cases / recursive acquisition of locks.

Comment (philipphofmann) [reply Sources/Swift/Integrations/AppHangTracking/SentryRunLoopDelayTracker.swift]: Yeah, to me, slight alarm bells are ringing because it looks quite fragile. It's quite easy to break and introduce deadlocks by mistake. Since we're using the SentryMutex here only for observers, which are infrequent, I wonder if the previous approach would be better, as the overhead of NSLock should be negligible compared to the SentryMutex. Or am I missing something, @philprime?

Comment (philipphofmann) [reply Sources/Swift/Integrations/AppHangTracking/SentryRunLoopDelayTracker.swift]: My bad. After an internal chat @philprime explained to me that we also had the `deinit` problem with NSLock, and the `_ = removed` makes sense no matter the locking. The merged change is legit.
